pkg/lib/manifest2po.py

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python3
+
+# Extracts translatable strings from manifest.json files.
+
+import argparse
+import json
+import os
+import re
+import sys
+
+import polib
+
+parser = argparse.ArgumentParser(
+    prog="manifest2po",
+    description="Extracts translatable strings from manifest.json files",
+)
+parser.add_argument("-d", "--directory", help="Base directory for input files")
+parser.add_argument("-o", "--output", help="Output file", required=True)
+parser.add_argument("files", nargs="+", help="One or more input files", metavar="FILE")
+
+args = parser.parse_args()
+
+entries = {}
+
+
+def process_manifest(filename, manifest):
+    if "menu" in manifest:
+        process_menu(filename, manifest["menu"])
+    if "tools" in manifest:
+        process_menu(filename, manifest["tools"])
+    if "bridges" in manifest:
+        process_bridges(filename, manifest["bridges"])
+
+
+def process_keywords(filename, keywords):
+    for v in keywords:
+        for keyword in v["matches"]:
+            push(msgid=keyword, location=(filename, "0"))
+
+
+def process_docs(filename, docs):
+    for doc in docs:
+        push(msgid=doc["label"], location=(filename, "0"))
+
+
+def process_menu(filename, menu):
+    for m in menu:
+        if "label" in menu[m]:
+            push(msgid=menu[m]["label"], location=(filename, "0"))
+        if "keywords" in menu[m]:
+            process_keywords(filename, menu[m]["keywords"])
+        if "docs" in menu[m]:
+            process_docs(filename, menu[m]["docs"])
+
+
+def process_bridges(filename, bridges):
+    for b in bridges:
+        if "label" in b:
+            push(msgid=b["label"], location=(filename, "0"))
+
+
+def push(msgid, location):
+    """
+    Push an entry onto the list
+    """
+    key = msgid
+    if key in entries:
+        entries[key].occurrences.append(location)
```

Comment:
I really dislike this global variable model.

I think it might make sense to switch to a generator model instead where you yield the msgids you encounter.  You can then collect those in the main function and assign the location to them (which would also allow you to avoid repeating `(filename "0")`.

From main you could maybe use something like a `defaultdict(set)` for building something like `dict[str, set[str]]` mapping msgids to filenames and from there write the pofile directly.

Reply:
To be a bit more clear, something like:

```python

def get_manifest_strings(manifest: Dict[str, object]) → Iterable[str]:
    yield from get_menu_strings(...)
    etc...

def main() -> None:
    ...

    strings = collections.defaultdict(set)
    for filename ...:
        manifest = parse...
        for msgid get_manifest_strings(manifest):
            strings[msgid].add(filename)

    ...write po file...
```

Reply:
hm while I can see the point of the iterable, IMHO it adds more complexity to a script that, in the end, it is easy.

in the js version also `filename` was a global variable, which i dropped in favour of passing it around; it isn't "that nice", however IMHO it is not that problematic either; I changed `entries` to be passed around as parameter